csharp/src/Ice/Proxy.cs

```diff
@@ -396,17 +397,21 @@ public static Connection GetConnection(this IObjectPrx prx)
                         catch (NoEndpointException ex)
                         {
                             // The reference has no endpoints or the previous retry policy asked to retry on a
-                            // different replica but no more replicas are available (in this case, we rethrow
-                            // the remote exception instead of the NoEndpointException).
-                            lastException = response == null ? ex : null;
+                            // different replica but no more replicas are available (in this case, we throw
+                            // the previous exception instead of the NoEndpointException).
+                            if (response == null && lastException == null)
+                            {
+                                lastException = ex;
+                            }
                             childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
                         }
                         catch (TransportException ex)
                         {
                             var closedException = ex as ConnectionClosedException;
-                            if (connection != null && closedException == null)
+                            connector ??= ex.Connector;
```

Comment:
We don't want to add the connector to the transport failures list, this `ConnectionClosedException` can be a normal condition, connection closed by ACM. We also use the closedException below to check if it was closed by the peer in which case it is safe to retry, we only retry idempotent or not sent TransportException otherwise independently of the retry policy.